Mathlib/InformationTheory/BinaryEntropy.lean

```diff
@@ -0,0 +1,345 @@
+
+@[simp] lemma deriv_one_minus (x : ℝ) : deriv (fun (y : ℝ) ↦ 1 - y) x = -1 := by
+  have onem (y : ℝ) : 1 - y = -(y + -1) := by ring
+  simp_rw [onem]
+  simp only [neg_add_rev, neg_neg, differentiableAt_const, deriv_const_add', deriv_neg'']
+
+@[simp] lemma differentiable_1_minusp (p : ℝ) : DifferentiableAt ℝ (fun p => 1 - p) p := by
+  have (p : ℝ) : 1 - p = -(p - 1) := by ring
+  simp_rw [this]
+  apply differentiableAt_neg_iff.mpr
+  apply DifferentiableAt.add_const
+  simp only [differentiableAt_id']
+
+-- TODO don't need assumptions
+lemma deriv_log_one_sub {x : ℝ} (hh : x ≠ 1): deriv (fun p ↦ log (1 - p)) x = -(1-x)⁻¹ := by
+  rw [deriv.log]
+  simp only [deriv_one_minus]
+  field_simp
+  exact differentiable_1_minusp x
+  exact sub_ne_zero.mpr hh.symm
+
+@[simp] lemma differentiableAt_log_const_neg {x c : ℝ} (h : x ≠ c) :
+    DifferentiableAt ℝ (fun p ↦ log (c - p)) x := by
+  apply DifferentiableAt.log
+  apply DifferentiableAt.sub
+  apply differentiableAt_const
+  apply differentiableAt_id'
+  exact sub_ne_zero.mpr (id (Ne.symm h))
+
+-- TODO don't need assumptions
+lemma deriv_h₂' {x : ℝ} (h: x ≠ 0) (hh : x ≠ 1) :
+    deriv (fun p => -p * log₂ p - (1 - p) * log₂ (1 - p)) x = log₂ (1 - x) - log₂ x := by
+  simp [log₂, logb]
+  rw [deriv_sub, deriv.neg]
+  simp_rw [mul_div]
+  rw [deriv_div_const, deriv_mul_log h, deriv_div_const]
+  simp_rw [mul_sub_right_distrib]
+  simp only [one_mul]
+  rw [deriv_sub, deriv_log_one_sub hh]
+  · rw [deriv_mul, deriv_id'']
+    rw [deriv.log]
+    simp only [one_mul, deriv_one_minus]
+    field_simp
+    ring_nf
+    calc -1 + (-log x - x * (1 - x)⁻¹) + (1 - x)⁻¹ + log (1 - x)
+      _ = -1 + (1 - x) * (1 - x)⁻¹ + log (1 - x) - log x := by ring
+      _ = -log x + log (1 - x) := by
+        field_simp [sub_ne_zero.mpr hh.symm]
+        ring
+    apply differentiable_1_minusp
+    exact sub_ne_zero.mpr hh.symm
+    apply differentiableAt_id'
+    exact differentiableAt_log_const_neg hh
+  · exact differentiableAt_log_const_neg hh
+  · apply DifferentiableAt.mul
+    apply differentiableAt_id'
+    apply DifferentiableAt.log
+    exact differentiable_1_minusp x
+    exact sub_ne_zero.mpr hh.symm
+  · apply DifferentiableAt.neg
+    apply DifferentiableAt.mul
+    exact differentiableAt_id'
+    apply DifferentiableAt.div_const
+    exact differentiableAt_log h
+  · apply DifferentiableAt.mul
+    apply differentiable_1_minusp
+    apply DifferentiableAt.div_const
+    exact differentiableAt_log_const_neg hh
+
+-- TODO don't need assumptions
+lemma deriv_h₂ {x : ℝ} (h: x ≠ 0) (hh : x ≠ 1) : deriv h₂ x = log₂ (1 - x) - log₂ x := by
+  unfold h₂
+  apply deriv_h₂' h hh
+
+/- Binary entropy has derivative `log₂ (1 - p) - log₂ p`. -/
+lemma hasDerivAt_h₂ {x : ℝ} (xne0: x ≠ 0) (gne1 : x ≠ 1) :
+    HasDerivAt h₂ (Entropy.h₂deriv x) x := by
+  have diffAtStuff : DifferentiableAt ℝ (fun p => -p * log₂ p - (1 - p) * log₂ (1 - p)) x := by
```

Comment:
Rename.

Reply:
Do you mean the lemma or the `have`? Assuming you mean the lemma: why rename it? Of course it has to be renamed once `h\_2` is renamed. For now, the pattern overall matches e.g. `Real.hasDerivAt_log`.

Reply:
A useful strategy is to run `fun_prop (disch:=trace_state;sorry)`. 

Then you look at the info view see what was sorried to close the goal. Now you can add these statement to the local context and use `fun_prop (disch:=assumption)`